File: rust/6.zigzag-conversion.rs
/*
 * @lc app=leetcode id=6 lang=rust
 *
 * [6] Zigzag Conversion
 */

// @lc code=start
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        if num_rows == 1 {
            return s;
        }
        let mut zigzag = String::new();
        let bytes = s.as_bytes();
        for row in 0..num_rows {
            let mut i = row as usize;
            let mut rev = false;
            while i < s.len() {
                zigzag.push(bytes[i] as char);
                let mut n = 0;
                if row == 0 || row == num_rows - 1 {
                    n = num_rows - 1;
                } else if rev {
                    n = row;
                } else {
                    n = num_rows - row - 1;
                }
                i += (2 * n) as usize;
                rev = !rev;
            }
        }
        return zigzag;
    }
}
// @lc code=end

File: javascript/6.zigzag-conversion.js
/*
 * @lc app=leetcode id=6 lang=javascript
 *
 * [6] Zigzag Conversion
 */

// @lc code=start
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
  if (numRows === 1) return s;

  const zigzag = [];
  for (let row = 0; row < numRows; row++) {
    let i = row;
    let rev = false;
    while (i < s.length) {
      zigzag.push(s[i]);
      let n = 0;
      if (row === 0 || row === numRows - 1) n = numRows - 1;
      else n = rev ? row : numRows - row - 1;
      i += 2 * n;
      rev = !rev;
    }
  }
  return zigzag.join("");
};
// @lc code=end
